src/ray/metrics/perf_counter.h

```diff
@@ -0,0 +1,142 @@
+namespace perf_counter {
+
+/// Include tag_def.h to define tag items
+#define RAY_TAG(name)                                                       \
+  static opencensus::tags::TagKey name##Key() {                             \
+    static const auto key = opencensus::tags::TagKey::Register(#name);      \
+    return key;                                                             \
+  }
+
+#include "tag_def.h"
+#undef RAY_TAG
+
+
+/// A thin wrapper that wraps the `opencensus::tag::measure` for using it simply.
+class Metric final {
+
+ public:
+  Metric(const std::string &name,
+         const std::string &description,
+         const std::string &unit) {
+    this->measure_ = std::make_shared<opencensus::stats::Measure<double>>(
+        opencensus::stats::Measure<double>::Register(
+            name, description, unit));
+  }
+
+  ~Metric() = default;
+
+  std::string GetName() const {
+    return measure_->GetDescriptor().name();
+  }
+
+  void Record(double value) {
+    Record(value, {});
+  }
+
+  void Record(double value, const std::vector<std::pair<opencensus::tags::TagKey::TagKey, std::string>>& tags) {
+    // global tags should be registered here.
+    static std::vector<std::pair<opencensus::tags::TagKey, std::string>> global_tags = {
+        {ray::perf_counter::JobNameKey(), "raylet"}
+    };
+
+    std::vector<std::pair<opencensus::tags::TagKey, std::string>> combined_tags(tags);
+    combined_tags.insert(std::end(combined_tags), std::begin(global_tags), std::end(global_tags));
+    opencensus::stats::Record({{*this->measure_, value}}, combined_tags);
+  }
+
+ private:
+  std::shared_ptr<opencensus::stats::Measure<double>> measure_ = nullptr;
+
+}; // class Metric
+
+/// Include metric_def.h to define tag items
+#define RAY_METRIC(name, description, unit)                                 \
+  static Metric name##_metric_ = Metric(std::string("raylet/") + #name,     \
+                                        description, unit);                 \
+  static Metric &name() {                                                   \
+    return name##_metric_;                                                  \
+  }
+
+#include "metrics_def.h"
+#undef RAY_METRIC
+
+
+/// The helper function for registering a view.
+  static void RegisterAsView(opencensus::stats::ViewDescriptor view_descriptor) {
+    opencensus::stats::View view(view_descriptor);
+    view_descriptor.RegisterForExport();
+  }
+
+
+/// The helper function for registering all views.
+  static void RegisterAllViews() {
+    {
+      const opencensus::stats::ViewDescriptor view_descriptor =
+          opencensus::stats::ViewDescriptor().set_name("raylet/latency")
+              .set_description("The latency of redis.")
+              .set_measure(RedisLatency().GetName())
+              .set_aggregation(opencensus::stats::Aggregation::Distribution(
+                  opencensus::stats::BucketBoundaries::Explicit({0, 25, 50, 75, 1000})))
+              .add_column(JobNameKey());
```

Comment:
@ericl  Another question I'm confused about yet:

We register many views right now, and generate a lot of duplicated code(like line 86 - 120).
Could we register some views in `Metric` to avoid the duplicated code, like this:

```C++
class Metric {
public:
  static std::shared_ptr<Metric> makeGauge(name, description, unit, tags) {
      auto metric = std::make_shared<Metric>(name, description, unit);
      
      const opencensus::stats::ViewDescriptor view_descriptor =
          opencensus::stats::ViewDescriptor().set_name(name)
              .set_description(description)
              .set_measure(metric.GetName())
              .set_aggregation(opencensus::stats::Aggregation::LastValue())
              .add_column(AllGlobalTags())
              .add_column(tags);
       RegisterAsView(view_descriptor);

      return ret;
  }

static std::shared_ptr<Metric> makeHistogram(name, description, unit, bucket_set, tags) {
    auto metric = std::make_shared<Metric>(name, description, unit);
    const opencensus::stats::ViewDescriptor view_descriptor =
          opencensus::stats::ViewDescriptor().set_name(name)
              .set_description(description)
              .set_measure(metric.GetName())
              .set_aggregation(opencensus::stats::Aggregation::Distribution(
                  opencensus::stats::BucketBoundaries::Explicit(bucket_set)))
              .add_column(AllGlobalTags())
              .add_column(tags));
       RegisterAsView(view_descriptor);
      return ret;
}

static std::shared_ptr<Metric> makeCount(name, description, unit, tags) {
// ...
}

static std::shared_ptr<Metric> makeSum(name, description, unit, tags) {
// ...
}
};
```

Then, we could use it in `perf_counter.h`:
```C++
namespace perf_counter {
std::shared_ptr<Metric> RedisLatency = Metric::makeHistogram("redis_latency", "xxx", "ms", {0, 100,500}, {JobNameKey(), NodeAddressKey()});
}
```

Also the name `makeGauge()` could be `makeLastValue()` and `makeHistogram()` could be `makeDistribution()` to be consistent with `opencensus naming`.

Reply:
That seems fine to me. Note that you don't need to have a ptr around Metric, it is a value type.